crates/sui-graphql-rpc/src/types/transaction_block/mod.rs

```diff
@@ -599,3 +509,150 @@ impl TryFrom<TransactionBlockEffects> for TransactionBlock {
+            if (after.is_scan_limited || scan_limit.is_some())
+                && after.tx_sequence_number > tx_bounds.lo
+            {
+                // When dealing with `scan_limit`, the first scanned transaction serves as the
+                // `startCursor`. This will always be 1 greater than the given `after` cursor.
+                let starting_tx_sequence_number = after.tx_sequence_number + 1;
+
+                // It is possible for the first scanned transaction to show up in the result set,
+                // and if so we should flip `is_scan_limited` to false.
+                let is_scan_limited = transactions.first().map_or(true, |tx| {
+                    tx.tx_sequence_number as u64 != starting_tx_sequence_number
+                });
+                let start_cursor = cursor::TransactionBlockCursor {
+                    checkpoint_viewed_at,
+                    tx_sequence_number: starting_tx_sequence_number,
+                    is_scan_limited,
+                };
+
+                conn.has_previous_page = true;
+                conn.start_cursor = Some(Cursor::new(start_cursor).encode_cursor());
+            }
+        }
+    }
+
+    if !conn.has_next_page && scan_limit.is_some() && tx_bounds.scan_has_next_page() {
+        // There are 4 scenarios that will yield `has_next_page=false`:
+        // 1. met `limit`, `paginate_results` doesn't detect more results + more to scan
+        // 2. met `limit`, `paginate_Results` doesn't detect more results + no more to scan
+        // 3. less than `limit`, `paginate_results` doesn't detect more results + more to scan
+        // 4. less than `limit`, `paginate_results` doesn't detect more results + no more to scan
+        // Regardless of the scenario, we can set the `endCursor` to the last transaction scanned.
+
+        let scan_hi = tx_bounds.scan_hi();
+
+        conn.has_next_page = true;
+        conn.end_cursor = Some(
+            Cursor::new(cursor::TransactionBlockCursor {
+                checkpoint_viewed_at,
+                tx_sequence_number: scan_hi,
+                is_scan_limited: transactions
+                    .last()
+                    .map_or(true, |tx| tx.tx_sequence_number as u64 != scan_hi),
+            })
+            .encode_cursor(),
+        );
+    }
+}
+
+fn apply_backward_scan_limited_pagination(
+    conn: &mut ScanConnection<String, TransactionBlock>,
+    page: &Page<Cursor>,
+    transactions: &Vec<StoredTransaction>,
+    tx_bounds: TxBounds,
+    scan_limit: Option<u64>,
+    checkpoint_viewed_at: u64,
+) {
+    if !conn.has_next_page {
+        if let Some(before) = page.before() {
+            if (before.is_scan_limited || scan_limit.is_some())
+                && before.tx_sequence_number < tx_bounds.hi
+            {
+                // When dealing with `scan_limit`, the last scanned transaction serves as the
+                // `endCursor`.
+                let ending_tx_sequence_number = before.tx_sequence_number - 1;
```

Comment:
Is there a risk that `tx_sequence_number` could be `0`? There's a similar question about whether in the forward case we can guarantee that we've validated that the `tx_sequence_number` is not `u64::max`.

Since these are user inputs, let's either include a `SAFETY` comment explaining why this is OK, or use `checked_` operations to make sure we don't accidentally underflow or overflow.

Reply:
yes, i modified the code to this to help elucidate what's going on, but didn't put any safety around the cursor-adjusting yet